l5kit/l5kit/simulation/unroll.py

```diff
@@ -0,0 +1,226 @@
+
+        simulated_outputs: List[SimulationOutputs] = []
+        for scene_idx in scene_indices:
+            simulated_outputs.append(SimulationOutputs(scene_idx, sim_dataset))
+        return simulated_outputs
+
+    @staticmethod
+    def update_agents(dataset: SimulationDataset, frame_idx: int, input_dict: Dict[str, torch.Tensor],
+                      output_dict: Dict[str, torch.Tensor]) -> None:
+        """Update the agents in frame_idx (across scenes) using agents_output_dict
+
+        :param dataset: the simulation dataset
+        :param frame_idx: index of the frame to modify
+        :param input_dict: the input to the agent model
+        :param output_dict: the output of the agent model
+        :return:
+        """
+
+        agents_update_dict: Dict[Tuple[int, int]: np.ndarray] = {}
+        input_dict = {k: v.cpu().numpy() for k, v in input_dict.items()}
+        output_dict = {k: v.cpu().numpy() for k, v in output_dict.items()}
+
+        world_from_agent = input_dict["world_from_agent"]
+        yaw = input_dict["yaw"]
+        pred_trs = transform_points(output_dict["positions"][:, :1], world_from_agent)[:, 0]
+        pred_yaws = yaw + output_dict["yaws"][:, 0]
+
+        next_agents = np.zeros(len(yaw), dtype=AGENT_DTYPE)
+        next_agents["centroid"] = pred_trs
+        next_agents["yaw"] = pred_yaws
+        next_agents["track_id"] = input_dict["track_id"]
+        next_agents["extent"] = input_dict["extent"]
+        next_agents["label_probabilities"][:, 3] = 1
+
+        for scene_idx, next_agent in zip(input_dict["scene_index"], next_agents):
+            agents_update_dict[(scene_idx, next_agent["track_id"])] = np.expand_dims(next_agent, 0)
+        dataset.set_agents(frame_idx, agents_update_dict)
+
+    @staticmethod
+    def update_ego(dataset: SimulationDataset, frame_idx: int, input_dict: Dict[str, torch.Tensor],
+                   output_dict: Dict[str, torch.Tensor]) -> None:
+        """Update ego across scenes for the given frame index.
+
+        :param dataset: The simulation dataset
+        :param frame_idx: index of the frame to modify
+        :param input_dict: the input to the ego model
+        :param output_dict: the output of the ego model
+        :return:
+        """
+        input_dict = {k: v.cpu().numpy() for k, v in input_dict.items()}
+        output_dict = {k: v.cpu().numpy() for k, v in output_dict.items()}
+
+        world_from_agent = input_dict["world_from_agent"]
+        yaw = input_dict["yaw"]
+        pred_trs = transform_points(output_dict["positions"][:, :1], world_from_agent)
+        pred_yaws = yaw + output_dict["yaws"][:, :1]
+
+        dataset.set_ego_for_frame(frame_idx, 0, pred_trs, pred_yaws)
+
+
+from l5kit.configs import load_config_data
+from l5kit.data import ChunkedDataset, LocalDataManager
+from l5kit.dataset import EgoDataset
+from l5kit.rasterization import build_rasterizer
```

Comment:
Move to the top

Reply:
this will be removed. it's here only for rapid testing